### PR TITLE
Add unique indexes and indexes for composite types

### PR DESCRIPTION
**Pull request by FelixMalfait**
Add support for indexes on composite fields and unicity constraint on indexes

This pull request includes several changes across multiple files to improve error handling, enforce unique constraints, and update database migrations. The most important changes include updating error messages for snack bars, adding a new command to enforce unique constraints, and updating database migrations to include new fields and constraints.

### Error Handling Improvements:
* [`packages/twenty-front/src/modules/error-handler/components/PromiseRejectionEffect.tsx`](diffhunk://#diff-e7dc05ced8e4730430f5c7fcd0c75b3aa723da438c26e0bef8130b614427dd9aL23-R23): Updated error messages in `enqueueSnackBar` to use `error.message` directly.
* [`packages/twenty-front/src/modules/object-metadata/hooks/useFindManyObjectMetadataItems.ts`](diffhunk://#diff-74c126d6bc7a5ed6b63be994d298df6669058034bfbc367b11045f9f31a3abe6L44-R46): Simplified error messages in `enqueueSnackBar`.
* [`packages/twenty-front/src/modules/object-record/hooks/useFindDuplicateRecords.ts`](diffhunk://#diff-af23a1d99639a66c251f87473e63e2b7bceaa4ee4f70fedfa0fcffe5c7d79181L56-R58): Simplified error messages in `enqueueSnackBar`.
* [`packages/twenty-front/src/modules/object-record/hooks/useHandleFindManyRecordsError.ts`](diffhunk://#diff-da04296cbe280202a1eaf6b1244a30490d4f400411bee139651172c59719088eL22-R24): Simplified error messages in `enqueueSnackBar`.

### New Command for Unique Constraints:
* [`packages/twenty-server/src/database/commands/upgrade-version/0-31/0-31-enforce-unique-constraints.command.ts`](diffhunk://#diff-8337096c8c80dd2619a5ba691ae5145101f8ae0368a75192a050047e8c6ab7cbR1-R159): Added a new command to enforce unique constraints on company domain names and person emails.
* [`packages/twenty-server/src/database/commands/upgrade-version/0-31/0-31-upgrade-version.command.ts`](diffhunk://#diff-20215e9981a53c7566e9cbff96715685125878f5bcb84fe461a7440f2e68f6fcR13-R14): Integrated the new `EnforceUniqueConstraintsCommand` into the upgrade process. [[1]](diffhunk://#diff-20215e9981a53c7566e9cbff96715685125878f5bcb84fe461a7440f2e68f6fcR13-R14) [[2]](diffhunk://#diff-20215e9981a53c7566e9cbff96715685125878f5bcb84fe461a7440f2e68f6fcR31) [[3]](diffhunk://#diff-20215e9981a53c7566e9cbff96715685125878f5bcb84fe461a7440f2e68f6fcR64-R68)
* [`packages/twenty-server/src/database/commands/upgrade-version/0-31/0-31-upgrade-version.module.ts`](diffhunk://#diff-da52814efc674c25ed55645f8ee2561013641a407f88423e705dd6c77b405527R7): Registered the new `EnforceUniqueConstraintsCommand` in the module. [[1]](diffhunk://#diff-da52814efc674c25ed55645f8ee2561013641a407f88423e705dd6c77b405527R7) [[2]](diffhunk://#diff-da52814efc674c25ed55645f8ee2561013641a407f88423e705dd6c77b405527R24)

### Database Migrations:
* [`packages/twenty-server/src/database/typeorm/metadata/migrations/1726757368824-migrationDebt.ts`](diffhunk://#diff-c450aeae7bc0ef4416a0ade2dc613ca3f688629f35d2a32f90a09c3f494febdcR1-R53): Added a migration to update the `relationMetadata_ondeleteaction_enum` and set default values.
* [`packages/twenty-server/src/database/typeorm/metadata/migrations/1726757368825-addIsUniqueToIndexMetadata.ts`](diffhunk://#diff-8f1e14bd7f6835ec2c3bb39bcc51e3c318a3008d576a981e682f4c985e746fbfR1-R19): Added a migration to include the `isUnique` field in `indexMetadata`.
* [`packages/twenty-server/src/database/typeorm/metadata/migrations/1726762935841-addCompostiveColumnToIndexFieldMetadata.ts`](diffhunk://#diff-7c96b7276c7722d41ff31de23b2de4d6e09adfdc74815356ba63bc96a2669440R1-R19): Added a migration to include the `compositeColumn` field in `indexFieldMetadata`.
* [`packages/twenty-server/src/database/typeorm/metadata/migrations/1726766871572-addWhereToIndexMetadata.ts`](diffhunk://#diff-26651295a975eb50e672dce0e4e274e861f66feb1b68105eee5a04df32796190R1-R14): Added a migration to include the `indexWhereClause` field in `indexMetadata`.

### GraphQL Exception Handling:
* [`packages/twenty-server/src/engine/api/graphql/workspace-query-runner/utils/workspace-query-runner-graphql-api-exception-handler.util.ts`](diffhunk://#diff-58445eb362dc89e31107777d39b592d7842d2ab09a223012ccd055da325270a8R1-R4): Enhanced exception handling for `QueryFailedError` to provide more specific error messages for unique constraint violations. [[1]](diffhunk://#diff-58445eb362dc89e31107777d39b592d7842d2ab09a223012ccd055da325270a8R1-R4) [[2]](diffhunk://#diff-58445eb362dc89e31107777d39b592d7842d2ab09a223012ccd055da325270a8R23-R59)
* [`packages/twenty-server/src/engine/api/graphql/workspace-resolver-builder/factories/create-many-resolver.factory.ts`](diffhunk://#diff-233d58ab2333586dd45e46e33d4f07e04a4b8adde4a11a48e25d86985e5a7943L58-R58): Updated the `workspaceQueryRunnerGraphqlApiExceptionHandler` call to include context.
* [`packages/twenty-server/src/engine/api/graphql/workspace-resolver-builder/factories/create-one-resolver.factory.ts`](diffhunk://#diff-68b803f0762c407f5d2d1f5f8d389655a60654a2dd2394a81318655dcd44dc43L58-R58): Updated the `workspaceQueryRunnerGraphqlApiExceptionHandler` call to include context.

### CODE REVIEW

**Comment by FelixMalfait** [comment]
Todo:
- check that having 2 null-s isn't considered as a unicity constraint violation
- update `GraphqlQueryUpdateManyResolverService.validate`, `GraphqlQueryUpdateOneResolverService.validate` and 
`GraphqlQueryCreateManyResolverService.deleteMany` to perform a pre-check there and throw a user-friendly unicity violation error at application level and not wait for the error to happen at the DB level (small negative impact on performance but that seems a better approach no @Weiko ?) 
- not important / after merging this, create an issue to see if we can learn from that unicity constraint error and build a mini DTO-framework like email-format validation or things like that in that validate method (?)

File: packages/twenty-front/src/modules/object-metadata/hooks/useFindManyObjectMetadataItems.ts [hunk 1]
```diff
@@ -41,12 +41,9 @@ export const useFindManyObjectMetadataItems = ({
     skip: skip || !apolloMetadataClient,
     onError: (error) => {
       logError('useFindManyObjectMetadataItems error : ' + error);
-      enqueueSnackBar(
-        `Error during useFindManyObjectMetadataItems, ${error.message}`,
-        {
-          variant: SnackBarVariant.Error,
-        },
-      );
+      enqueueSnackBar(`${error.message}`, {
```
Comment (FelixMalfait) [inline packages/twenty-front/src/modules/object-metadata/hooks/useFindManyObjectMetadataItems.ts:44]: was too dev-oriented

File: packages/twenty-front/src/modules/object-record/hooks/useFindDuplicateRecords.ts [hunk 2]
```diff
@@ -53,12 +53,9 @@ export const useFindDuplicateRecords = <T extends ObjectRecord = ObjectRecord>({
             `useFindDuplicateRecords for "${objectMetadataItem.nameSingular}" error : ` +
               error,
           );
-          enqueueSnackBar(
-            `Error during useFindDuplicateRecords for "${objectMetadataItem.nameSingular}", ${error.message}`,
-            {
-              variant: SnackBarVariant.Error,
-            },
-          );
+          enqueueSnackBar(`Error finding duplicates:", ${error.message}`, {
```
Comment (FelixMalfait) [inline packages/twenty-front/src/modules/object-record/hooks/useFindDuplicateRecords.ts:56]: was too dev-oriented

File: packages/twenty-front/src/modules/error-handler/components/PromiseRejectionEffect.tsx [hunk 3]
```diff
@@ -20,7 +20,7 @@ export const PromiseRejectionEffect = () => {
           },
         );
       } else {
-        enqueueSnackBar(`Error: ${event.reason}`, {
+        enqueueSnackBar(`${error.message}`, {
```
Comment (FelixMalfait) [inline packages/twenty-front/src/modules/error-handler/components/PromiseRejectionEffect.tsx:23]: red color and message usually already convey the fact that it's an error, and there's limited space

File: packages/twenty-server/src/database/typeorm/metadata/migrations/1726757368824-migrationDebt.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class MigrationDebt1726757368824 implements MigrationInterface {
```
Comment (FelixMalfait) [inline packages/twenty-server/src/database/typeorm/metadata/migrations/1726757368824-migrationDebt.ts:3]: This is not directly related to this PR but it reflects that migrations <> typeorm entities were out of sync. I ran this a few weeks ago when opening the PR but it seems the debt hasn't been picked up by anyone since

File: packages/twenty-server/src/database/typeorm/metadata/migrations/1726762935841-addColumnNameToIndexFieldMetadata.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class AddColumnNameToIndexFieldMetadata1726762935841
```
Comment (FelixMalfait) [inline packages/twenty-server/src/database/typeorm/metadata/migrations/1726762935841-addColumnNameToIndexFieldMetadata.ts:3]: Migration name doesn't reflect well what's inside. 
Todo: rename

File: packages/twenty-server/src/database/commands/upgrade-version/0-31/0-31-enforce-unique-constraints.command.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+import {
+  ActiveWorkspacesCommandOptions,
+  ActiveWorkspacesCommandRunner,
+} from 'src/database/commands/active-workspaces.command';
+import { Workspace } from 'src/engine/core-modules/workspace/workspace.entity';
+import { TwentyORMGlobalManager } from 'src/engine/twenty-orm/twenty-orm-global.manager';
+
+@Command({
+  name: 'upgrade-0.31:enforce-unique-constraints',
```
Comment (FelixMalfait) [inline packages/twenty-server/src/database/commands/upgrade-version/0-31/0-31-enforce-unique-constraints.command.ts:15]: This isn't ready to be reviewed yet

File: packages/twenty-server/src/database/typeorm/metadata/migrations/1726757368824-migrationDebt.ts (hunk 4 above)
Comment (Weiko) [reply packages/twenty-server/src/database/typeorm/metadata/migrations/1726757368824-migrationDebt.ts:3]: Thanks! Ideally we should enforce a rule (danger?)
If *.entity.ts files are updated, a migration should be generated 🤔

File: packages/twenty-server/src/database/typeorm/metadata/migrations/1726762935841-addCompostiveColumnToIndexFieldMetadata.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class AddCompositeColumnToIndexFieldMetadata1726762935841
+  implements MigrationInterface
+{
+  name = 'AddCompositeColumnToIndexFieldMetadata1726762935841';
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      `ALTER TABLE "metadata"."indexFieldMetadata" ADD "compositeColumn" character varying`,
```
Comment (Weiko) [inline packages/twenty-server/src/database/typeorm/metadata/migrations/1726762935841-addCompostiveColumnToIndexFieldMetadata.ts:10]: Why do we need compositeColumn here? Should be reflected by the fact that indexMetadata has a oneToMany indexFieldMetadata or maybe I misunderstood this column?

File: packages/twenty-server/src/engine/api/graphql/workspace-query-runner/utils/workspace-query-runner-graphql-api-exception-handler.util.ts [hunk 8]
```diff
@@ -16,7 +20,43 @@ import {
...
+
+        throw new UserInputError(
+          `A duplicate entry was detected. The combination of ${columnNames} must be unique.`,
+        );
+      }
+    }
+
+    // If it's not a duplicate key error, or we couldn't parse the index name,
+    // we'll throw a generic error
+    throw new UserInputError(
```
Comment (Weiko) [inline packages/twenty-server/src/engine/api/graphql/workspace-query-runner/utils/workspace-query-runner-graphql-api-exception-handler.util.ts:55]: This won't be captured by sentry, do we want that?

File: packages/twenty-server/src/engine/twenty-orm/decorators/workspace-column-index.decorator.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
+import { generateDeterministicIndexName } from 'src/engine/metadata-modules/index-metadata/utils/generate-deterministic-index-name';
+import { WorkspaceIndexOptions } from 'src/engine/twenty-orm/decorators/workspace-index.decorator';
+import { metadataArgsStorage } from 'src/engine/twenty-orm/storage/metadata-args.storage';
+import { getColumnsForIndex } from 'src/engine/twenty-orm/utils/get-default-columns-for-index.util';
+import { convertClassNameToObjectMetadataName } from 'src/engine/workspace-manager/workspace-sync-metadata/utils/convert-class-to-object-metadata-name.util';
+import { TypedReflect } from 'src/utils/typed-reflect';
+
+export function WorkspaceColumnIndex(
```
Comment (Weiko) [inline packages/twenty-server/src/engine/twenty-orm/decorators/workspace-column-index.decorator.ts:8]: Why do we need 2 decorators now? You think the code would have been to hard to read with both use cases?

File: packages/twenty-front/src/modules/object-metadata/components/ObjectMetadataItemsLoadEffect.tsx [hunk 10]
```diff
@@ -6,24 +6,11 @@ import { currentUserState } from '@/auth/states/currentUserState';
...
 import { useFindManyObjectMetadataItems } from '@/object-metadata/hooks/useFindManyObjectMetadataItems';
 import { objectMetadataItemsState } from '@/object-metadata/states/objectMetadataItemsState';
-import { ObjectMetadataItem } from '@/object-metadata/types/ObjectMetadataItem';
-import { FieldMetadataType } from '~/generated-metadata/graphql';
 import { WorkspaceActivationStatus } from '~/generated/graphql';
 import { generatedMockObjectMetadataItems } from '~/testing/mock-data/generatedMockObjectMetadataItems';
 import { isDeeplyEqual } from '~/utils/isDeeplyEqual';
 import { isUndefinedOrNull } from '~/utils/isUndefinedOrNull';
 
-const filterTsVectorFields = (
```
[the remark this answers is not in the document]
Comment (Weiko) [reply packages/twenty-front/src/modules/object-metadata/components/ObjectMetadataItemsLoadEffect.tsx:16]: Oh I see it seems you handled it differently 👍

File: packages/twenty-server/src/engine/api/graphql/workspace-resolver-builder/factories/create-many-resolver.factory.ts [hunk 11]
```diff
@@ -40,7 +40,7 @@ export class CreateManyResolverFactory
 
         return await this.graphqlQueryRunnerService.createMany(args, options);
       } catch (error) {
-        workspaceQueryRunnerGraphqlApiExceptionHandler(error);
+        workspaceQueryRunnerGraphqlApiExceptionHandler(error, context);
```
Comment (Weiko) [inline packages/twenty-server/src/engine/api/graphql/workspace-resolver-builder/factories/create-many-resolver.factory.ts:43]: should we use internalContext here as well to keep it consistent and avoid issues with _context?

File: packages/twenty-server/src/engine/api/graphql/workspace-schema-builder/services/type-mapper.service.ts [hunk 12]
```diff
@@ -114,6 +115,7 @@ export class TypeMapperService {
       [FieldMetadataType.RAW_JSON, RawJsonFilterType],
       [FieldMetadataType.RICH_TEXT, StringFilterType],
       [FieldMetadataType.ARRAY, ArrayFilterType],
+      [FieldMetadataType.TS_VECTOR, StringFilterType],
```
Comment (Weiko) [inline packages/twenty-server/src/engine/api/graphql/workspace-schema-builder/services/type-mapper.service.ts:118]: I'm not sure we want that already. TS_VECTOR is dedicated to searchVector field and search is only applied in the search resolver (which does not contain filter). We might want to combine filters later on but I'm not sure this change is needed? This won't be properly reflected in the API at least, your filter won't work as expected with a StringFilter on a tsvector column

**Review by charlesBochet — APPROVED** [review]
Big one! Let's iterate from there!
